### PR TITLE
refactor(Location): out of router and into platform/common

### PR DESCRIPTION
**Pull request by NathanWalker**
- [x] update documentation to reflect changes
- [x] discuss barrels/decide final placement for several auxiliary location classes.
- [x] update `public_api_spec.ts` with new api changes
- [x] squash and update commit message to reflect breaking changes

@mhevery I don't want this to affect the hard work going into making the ~ 10 KB goal for core.

I could foresee `Location`, `LocationStrategy`, `HashLocationStrategy`, `PathLocationStrategy`, `APP_BASE_HREF` going into `angular2/common`.

Please review when possible and let me know your thoughts.

### CODE REVIEW

**Comment by NathanWalker** [comment]
Ok. I just added 1 more commit that provides a 2nd alternative refactoring of `Location` into `common`. 
You can now compare both 1st and 2nd commits to see both refactor paths.

After working with the first refactor pass into core (the 1st commit), it occurred to me that a `Location` service (IMO) is perhaps, not a `core` feature. There are many applications that never need a `Location` service however, it is **common**. This is also advantageous as it does not affect `core`.

For this reason, I wanted to provide an alternative refactor which instead wraps `Location` into `angular2/common`. With the 2nd commit, the breaking changes would be:

BEFORE:
```
import {
  PlatformLocation,
  Location,
  LocationStrategy,
  HashLocationStrategy,
  PathLocationStrategy,
  APP_BASE_HREF}
from 'angular2/router';
```

AFTER:
```
import {PlatformLocation} from 'angular2/src/facade/platform_location';
import {
  Location,
  LocationStrategy,
  HashLocationStrategy,
  PathLocationStrategy,
  APP_BASE_HREF}
from 'angular2/common';
```

Since `PlatformLocation` should not be used directly by a common application developer (more for advanced usage), I believe this is well served in `facade` and not provided by a barrel. 

Please let me know which path you like better and I'll squash/add breaking changes to commit message, then update the documentation throughout the files to appropriately reflect the changes as well as update  `public_api_spec.ts` per your approval.

CI breaking is due only to `public_api_spec.ts` regarding api changes here.

**Comment by gdi2290** [comment]
you have to wait for this to be merged https://github.com/angular/angular/pull/7984

**Comment by gdi2290** [comment]
we also need to merge History class before refactoring the router. At the moment the current Location handles both Location and History

**Comment by NathanWalker** [comment]
Thanks for heads up @gdi2290, is there an issue or PR for merging History class?
I could create one?

Also I would greatly appreciate your feedback on this PR.
Main goal is to make Location service available without including the router package.
I'm open to any refactor paths to achieve that; the 2 here are just a way forward that appears clean and makes sense to me.

**Comment by mhevery** [comment]
Hi @NathanWalker, thanks for your PR. I don't think that `Location` should be part of the `core` since there could be platforms which do not have location concept. Also different platform may have different shape of `Location`, for example stack or a tree, hence it is hard to generalize. 

I could see how you may want to share the `Location` between `Browser` and `Server` but I don

**Comment by NathanWalker** [comment]
Agreed. On my 2nd commit I show an alternate refactor path placing it in `angular2/common` instead, however I assume that still ends up in the core barrel (bundled in `angular2.dev.js`)?

Would it make sense to have a separate barrel for this, perhaps `angular2/location` which would get bundled to 'angular2/bundles/location.js'?

Just trying to create a path forward for those trying to upgrade from Angular 1.x which provided `$location` as a core service. My company's large Angular 1.x app for instance has an **un**authenticated portion of our app which does *not* use the `router` however uses `$location` extensively for various marketing reasons. Therefore, we are in process of upgrading to 2.0 and facing the scenario of having to include the `router` bundle of 76 KB in those **un**authenticated areas just to get access to the `Location` service :(

**Comment by mhevery** [comment]
I think `Location` should be moved to `src/platform/browser` or `src/platform/common`. If you could move it there I would be fine with it. Could you update the PR and squash it?

**Comment by NathanWalker** [comment]
I moved it into `src/platform/browser`. I like it; makes a lot of sense to me.
Squashed commit and added breaking change block to message.

I'll update the documentation throughout the src by Monday/Tuesday. In meantime, when you can, please review these changes and let me know if you'd like anything else changed. Thanks for your attention to this.

File: modules/angular2/src/router/directives/router_link.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
 import {Directive} from 'angular2/core';
+import {Location} from 'angular2/src/platform/browser/location';
```
Comment (mhevery) [inline modules/angular2/src/router/directives/router_link.ts]: In general one should never import anything between libraries which includes `src` in the URL. You can think of Angular2 and Router as separate libraries.

The correct import should be  from `angular2/platform/...` Which means that the `Location` and friends need to be re-exported from there. Think of `src` as private and `angular2/platform` as public. Library can only import from another library public API.

This is a general issue I have seen in many imports. Please canonicalize them.

File: modules/angular2/test/public_api_spec.ts [hunk 2]
```diff
@@ -284,18 +284,27 @@ var NG_INSTRUMENTATION =
...
+  'LocationStrategy',
+  'PathLocationStrategy',
+  'PlatformLocation',
   'Title',
   'bootstrap',
   'disableDebugTools',
   'enableDebugTools',
-  'inspectNativeElement'
+  'inspectNativeElement',
+  'joinWithSlash',
```
Comment (mhevery) [inline modules/angular2/test/public_api_spec.ts]: 'joinWithSlash` and `normilzeQueryParams` should be a static member on existing type such as `Location`. I don't think they should be top level imports.

File: modules/angular2/test/router/directives/router_link_spec.ts [hunk 3]
```diff
@@ -17,10 +17,10 @@ import {
 import {SpyRouter, SpyLocation} from '../spies';
 
 import {provide, Component} from 'angular2/core';
+import {Location} from 'angular2/platform/browser';
```
Comment (mhevery) [inline modules/angular2/test/router/directives/router_link_spec.ts]: correct import

**Comment by mhevery** [comment]
Other than inconsistent import's this looks good to me. Please ping when ready so that we can merge it.

File: modules/angular2/test/public_api_spec.ts [hunk 4]
```diff
@@ -165,6 +165,11 @@ var NG_PLATFORM_BROWSER = [
   'enableDebugTools', 'inspectNativeElement'
 ];
 
+var NG_PLATFORM_COMMON = [
+  'APP_BASE_HREF', 'HashLocationStrategy', 'Location', 'LocationStrategy', 'PathLocationStrategy',
+  'PlatformLocation'
+];
+
```
Comment (NathanWalker) [inline modules/angular2/test/public_api_spec.ts]: May not be necessary but I thought you may want to have an API check for this to help with further testing if more is added to `platform/common` in the future.

**Comment by NathanWalker** [comment]
Ok due to fact that many tests can be run outside the browser, I opted to go with your other suggestion `platform/common` which works out the best.

Using `platform/browser` was problematic as it would execute imports trailing down to `angular2/src/platform/browser/tools/common_tools` which contains:
```
import {window} from 'angular2/src/facade/browser';
```

And this was not good for platform agnostic configurations and tests, etc. I thought about placing them in `platform/common_dom` however I think that would probably cause confusion down the line as they are not `DOM` specific.

`platform/common` was a great suggestion you had and this also provides a place for other platform agnostic services to grow into the future.

Please have a look when you can and hopefully this still meets your approval.

**Comment by NathanWalker** [comment]
I should mention I did not intend to change direction after the PR was flagged 'cleanup' however when working through the failing tests, the above emerged as the best solution.

**Comment by NathanWalker** [comment]
Just for ref (mentioned in commit msg):

`Location` and other related providers have been moved out of `router` and into `platform/common`. `BrowserPlatformLocation` is not meant to be used directly however advanced configurations may use it via the following import change.

Before:

```
import {
  PlatformLocation,
  Location,
  LocationStrategy,
  HashLocationStrategy,
  PathLocationStrategy,
  APP_BASE_HREF}
from 'angular2/router';

import {BrowserPlatformLocation} from 'angular2/src/router/location/browser_platform_location';
```

After:

```
import {
  PlatformLocation,
  Location,
  LocationStrategy,
  HashLocationStrategy,
  PathLocationStrategy,
  APP_BASE_HREF}
from 'angular2/platform/common';

import {BrowserPlatformLocation} from 'angular2/src/platform/browser/location/browser_platform_location';
```

**Comment by mhevery** [comment]
Looking good. You have a broken test:

```
	Expected [ '-UrlChangeEvent', '-UrlChangeListener' ] to equal [  ].
	main/</</<@/home/travis/build/angular/angular/dist/js/dev/es5/angular2/test/public_api_spec.js:200:17 <- diffing_plugin_wrapper-output_path-V43qaiXg.tmp/angular2/test/public_api_spec.ts:213:8
	_it/<@/home/travis/build/angular/angular/dist/js/dev/es5/angular2/src/testing/testing_internal.js:170:17 <- diffing_plugin_wrapper-output_path-V43qaiXg.tmp/angular2/src/testing/testing_internal.ts:171:21
```

Can you have a look at it? Otherwise this is good to go.

**Comment by NathanWalker** [comment]
Whenever I try to fix it for Dart it breaks for JS and vice versa.
Fixing the above error for JS (in saucelabs) then creates this Dart error which is the opposite:
```
[19:45:52] Starting 'test.unit.dart/ci'...
[19:45:52] Starting 'test.dart.dartium_symlink'...
[19:45:52] Finished 'test.dart.dartium_symlink' after 23 ms
[19:45:52] Starting '!test.unit.dart/run/angular2'...
00:24 +222 ~1: Form ... valueChanges should not fire an event when an excluded control is updated  
  Skip: skipping xit
00:47 +1596 ~2: Missing directive checks                                                           
  Skip: skipping xdescribe
00:51 +1724 ~3: NgZone ... run should run subscriber listeners in the subscription zone (outside)  
  Skip: skipping xit
00:51 +1755 ~4: NgZone ... run should run subscriber listeners in the subscription zone (outside)  
  Skip: skipping xit
00:56 +1952 ~4 -1: public API should fail if public API for ngPlatformCommon has changed           
  Expected: has different properties
    Actual: ['+UrlChangeEvent', '+UrlChangeListener']
     Which: is equal to [+UrlChangeEvent, +UrlChangeListener]. Expected: []
```

It appears Dart expects them whereas JS (saucelabs) does not:
```
[00:46:11] Starting 'test.unit.js.sauce/ci'...
...
Firefox 42.0.0 (Linux 0.0.0) public API should fail if public API for ngPlatformCommon has changed FAILED
	Expected [ '-UrlChangeEvent', '-UrlChangeListener' ] to equal [  ].
	main/</</<@/home/travis/build/angular/angular/dist/js/dev/es5/angular2/test/public_api_spec.js:200:17
```

Am I missing something?  Maybe a separate Dart file I'm not seeing?

**Comment by NathanWalker** [comment]
I've tried a couple things but unfortunately am stumped on the final CI failure related to `Dart` and testing `platform/common` public api. Following the consistency of the setup in `symbol_inspector.dart`, I would think this would be valid:
```
import 'dart:mirrors';
import './simple_library.dart' as simple_library;
import 'package:angular2/common.dart';
import 'package:angular2/compiler.dart';
import 'package:angular2/core.dart';
import 'package:angular2/instrumentation.dart';
import 'package:angular2/platform/browser.dart';
import 'package:angular2/platform/common.dart';  <---

...
const LIB_MAP = const {
  'simple_library': 'angular2.test.symbol_inspector.simple_library',
  'ngCommon': 'angular2.common',
  'ngCompiler': 'angular2.compiler',
  'ngCore': 'angular2.core',
  'ngInstrumentation': 'angular2.instrumentation',
  'ngPlatformBrowser': 'angular2.platform.browser',
  'ngPlatformCommon': 'angular2.platform.common' <---
};
```

However, this results:
```
dist/dart/angular2:[INFO] type: HINT Unused import (/home/travis/build/angular/angular/dist/dart/angular2/test/symbol_inspector/symbol_inspector.dart, line 10, col 8)
import 'package:angular2/platform/common.dart';
```

I tried removing the `import` however that results in errors related to `ngPlatformCommon` and unknown `angular2.platform.common` which I would have thought the import would solve.

If I remove this `'ngPlatformCommon': 'angular2.platform.common'` altogether, it results in errors when it checks the public api. 

I think I'm confused on a build step needed for `Dart` vs. `JS`. Still brainstorming as to how to solve but if anyone knows, please let me know.

**Comment by NathanWalker** [comment]
Whenever I try to fix it for Dart it breaks for JS and vice versa.
Fixing the above error for JS (in saucelabs) then creates this Dart error which is the opposite:
```
[19:45:52] Starting 'test.unit.dart/ci'...
[19:45:52] Starting 'test.dart.dartium_symlink'...
[19:45:52] Finished 'test.dart.dartium_symlink' after 23 ms
[19:45:52] Starting '!test.unit.dart/run/angular2'...
...
00:56 +1952 ~4 -1: public API should fail if public API for ngPlatformCommon has changed           
  Expected: has different properties
    Actual: ['+UrlChangeEvent', '+UrlChangeListener']
     Which: is equal to [+UrlChangeEvent, +UrlChangeListener]. Expected: []
```

It appears Dart expects them whereas JS (saucelabs) does not:
```
[00:46:11] Starting 'test.unit.js.sauce/ci'...
...
Firefox 42.0.0 (Linux 0.0.0) public API should fail if public API for ngPlatformCommon has changed FAILED
	Expected [ '-UrlChangeEvent', '-UrlChangeListener' ] to equal [  ].
	main/</</<@/home/travis/build/angular/angular/dist/js/dev/es5/angular2/test/public_api_spec.js:200:17
```

Am I missing something?  Maybe a separate Dart file I'm not seeing?

**Comment by mhevery** [comment]
I think because `UrlChangeEvent` is an interface it only shows up in Dart. Adding `UrlChangeEvent:dart` should fix the problem.

**Comment by NathanWalker** [comment]
You're brilliant @mhevery. Don't let nobody tell ya' different.
That got it, thanks!

**Comment by NathanWalker** [comment]
Just rebased latest master. Everything still ok, I think CI has a transient error (?)...
```
module.js:327
    throw err;
    ^
Error: Cannot find module 'fs-extra'
```

screen shot 2016-04-14 at 7 59 18 am

**Comment by mhevery** [comment]
Thank you. Added the right labels to get it merged...

**Comment by Delagen** [comment]
+1 #8240

**Pull request closed without merging**